cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/SupportBundleHttpHandler.java

```diff
@@ -0,0 +1,568 @@
+              applicationPair.getValue().getConfiguration(),
+              latestRunRecordDetail != null ? latestRunRecordDetail.getStartTs() : 0,
+              latestRunRecordDetail != null && latestRunRecordDetail.getStopTs() != null
+                ? latestRunRecordDetail.getStopTs()
+                : DateTime.now().getMillis());
+          generateLogFileAndRunInfo(
+            latestProgramRunId, latestRunRecordDetail, appFolderPath.toString(), metrics);
+        }
+      }
+    } catch (Exception e) {
+      LOG.error("Can not generate support bundle: ", e);
+    } finally {
+      try {
+        if (statusJson != null) {
+          statusJson.close();
+        }
+      } catch (IOException e) {
+        LOG.error("Can not close status json file ", e);
+      }
+    }
+  }
+
+  /** Gets oldest folder from the root directory */
+  private File getOldestFolder(File baseDirectory) {
+    File[] supportFiles = baseDirectory.listFiles();
+    long oldestDate = Long.MAX_VALUE;
+    File oldestFile = null;
+    if (supportFiles != null && supportFiles.length > 0) {
+      for (File f : supportFiles) {
+        if (f.lastModified() < oldestDate) {
+          oldestDate = f.lastModified();
+          oldestFile = f;
+        }
+      }
+    }
+    return oldestFile;
+  }
+
+  /** Deletes old folders after certain number of folders exist */
+  private void deleteOldFolders(File oldFilesDirectory) {
+    String[] entries = oldFilesDirectory.list();
+    if (entries != null && entries.length > 0) {
+      for (String s : entries) {
+        File currentFile = new File(oldFilesDirectory.getPath(), s);
+        // Recursive the full directory and delete all old files
+        if (currentFile.isDirectory()) {
+          deleteOldFolders(currentFile);
+        } else {
+          currentFile.delete();
+        }
+      }
+    }
+    oldFilesDirectory.delete();
+  }
+
+  /** Generates system log */
+  private void generateSystemLog(String systemLogPath, File basePath, List<ApplicationRecord> apps)
+    throws Exception {
+    String componentId = "services";
+    for (String serviceId : serviceList) {
+      CompletableFuture<Void> futureService =
+        CompletableFuture.runAsync(
+          () -> {
+            if (serviceId.equals("pipeline")) {
+              generateLogFile(
+                LoggingContextHelper.getLoggingContext(
+                  "system",
+                  serviceId,
+                  "studio",
+                  ProgramType.valueOfCategoryName(componentId)),
+                systemLogPath,
+                serviceId + "-system-log.txt");
+            } else {
+              generateLogFile(
+                LoggingContextHelper.getLoggingContext(
+                  Id.Namespace.SYSTEM.getId(), componentId, serviceId),
+                systemLogPath,
+                serviceId + "-system-log.txt");
+            }
+          });
+      futureService.get();
+      if (apps == null || apps.size() == 0) {
```

Comment:
Why `apps` is null? When you take a collection, always better to use empty instead of null to denote nothing.

Reply:
Also, use `apps.isEmpty` instead of `apps.size() == 0`.